bindings/mssql/mssql.go

```diff
@@ -0,0 +1,158 @@
+package mssql
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/logger"
+	_ "github.com/denisenkom/go-mssqldb"
+	"github.com/pkg/errors"
+)
+
+const (
+	// list of operations.
+	execOperation  bindings.OperationKind = "exec"
+	queryOperation bindings.OperationKind = "query"
+	closeOperation bindings.OperationKind = "close"
+
+	// configurations to connect to MSSQL, either a data source name represent by URL.
+	connectionURLKey = "url"
+
+	// keys from request's metadata.
+	commandSQLKey = "sql"
+)
+
+type MSSQL struct {
+	db     *sql.DB
+	logger logger.Logger
+}
+
+var _ = bindings.OutputBinding(&MSSQL{})
+
+
+// NewMSSQL returns a new MSSQL output binding.
+func NewMSSQL(logger logger.Logger) *MSSQL {
+	return &MSSQL{logger: logger}
+}
+
+func (m *MSSQL) Init(metadata bindings.Metadata) error {
+	m.logger.Debug("Initializing MSSQL binding")
+
+	p := metadata.Properties
+	url, ok := p[connectionURLKey]
+	if !ok || url == "" {
+		return fmt.Errorf("missing MSSQL connection string")
+	}
+
+	db, err := sql.Open("sqlserver", url)
+	if err != nil {
+		return fmt.Errorf("%q", err)
+	}
+	m.db = db
+	return nil
+}
+
+func (m *MSSQL) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	if req == nil {
+		return nil, errors.Errorf("invoke request required")
+	}
+
+	if req.Operation == closeOperation {
+		return nil, m.db.Close()
+	}
+
+	if req.Metadata == nil {
+		return nil, errors.Errorf("metadata required")
+	}
+	m.logger.Debugf("operation: %v", req.Operation)
+
+	s, ok := req.Metadata[commandSQLKey]
+	if !ok || s == "" {
+		return nil, errors.Errorf("required metadata not set: %s", commandSQLKey)
+	}
+
+	startTime := time.Now().UTC()
+
+	resp := &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			respOpKey:        string(req.Operation),
+			respSQLKey:       s,
+			respStartTimeKey: startTime.Format(time.RFC3339Nano),
+		},
+	}
+	switch req.Operation {

```

Comment:
You didn't implement the exec operation!

Also, when implementing that - please implement it using Prepared Statements. The request should additionally include the arguments to be sent to the prepared statement. See this issue for reference https://github.com/dapr/components-contrib/issues/1699

Given this is a new binding, we should not even allow anything that is not a prepared statement!

Reply:
This example is from Postgres -- not sure if a similar format exists for SQLServer. If it does, we should use prepared statements!

```
{
  "operation": "exec",
  "metadata": {
    "sql": "INSERT INTO foo (id, c1, ts) VALUES ($1, $2, $3)",
    "args": [
        1, 
        'demo', 
        '2020-09-24T11:45:05Z07:00'
    ]
  }
}
```